android-components/components/feature/app-links/src/main/java/mozilla/components/feature/app/links/AppLinksUseCases.kt

```diff
@@ -295,11 +295,41 @@ class AppLinksUseCases(
         @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
         internal var redirectCache: AppLinkRedirectCache? = null
 
+        @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+        internal var userDoNotInterceptCache: MutableMap<Int, Long> = mutableMapOf()
+
         @VisibleForTesting
         internal fun clearRedirectCache() {
             redirectCache = null
         }
 
+        @VisibleForTesting
+        internal fun getCacheKey(url: String, appIntent: Intent?): Int? {
+            return Uri.parse(url)?.let { uri ->
+                when {
+                    appIntent?.component?.packageName != null -> appIntent.component?.packageName
+                    !uri.isHttpOrHttps -> uri.scheme
+                    else -> uri.host // worst case we do not prompt again on this host
+                }.hashCode()
+            }
+        }
+
+        internal fun inUserDoNotIntercept(url: String, appIntent: Intent?): Boolean {
+            val cacheKey = getCacheKey(url, appIntent)
+            val cacheTimeStamp = userDoNotInterceptCache[cacheKey]
+            val currentTimeStamp = SystemClock.elapsedRealtime()
+
+            return cacheTimeStamp != null &&
+                currentTimeStamp <= (cacheTimeStamp + APP_LINKS_DO_NOT_OPEN_CACHE_INTERVAL)
+        }
+
+        internal fun addUserDoNotIntercept(url: String, appIntent: Intent?) {
+            val cacheKey = getCacheKey(url, appIntent)
+            cacheKey?.let {
+                userDoNotInterceptCache[it] = SystemClock.elapsedRealtime()
+            }
+        }
```

Comment:
I think those functions can live directly into AppLinksUseCases instead of the Companion object, and be called from `AppLinksFeatures` and `AppLinksInterceptor` using the instance of `appLinksUseCases` they already have. What do you think? 

Reply:
Unfortunately `userDoNotInterceptCache` needs to be in companion.  Since `AppLinksFeature` and `AppLinksInterceptor` doesn't use the same instance of `AppLinksUseCases`.  So moving these functions outside of the companion might be confusing, since we don't need an instance of `AppLinksUseCases` to access the cache.  We should allow the user to access it without creating an instance.

Reply:
As discussed, I've moved the cache into `AppLinksInterceptor` . Thanks giving me this idea!